airflow/providers/microsoft/azure/operators/container_instances.py

```diff
@@ -86,6 +86,12 @@ class AzureContainerInstancesOperator(BaseOperator):
     :param container_timeout: max time allowed for the execution of
         the container instance.
     :param tags: azure tags as dict of str:str
+    :param xcom_all: Control if logs are pushed to XCOM similarly to how DockerOperator does.
```

Comment:
I'm not sure what that means. 

To me, as an Airflow newbie, it seemed reasonable to provide the same behavior as the bash operator and the docker operator. I didn't want to mess with the existing return_value, so I put it under logs, but fundamentally I'm trying to get a return value out of the last line of the logs.

How would "adding operator extra link to remote resource" help? Or can you point me to an example? I'm a bit confused as to what this would look like. Thanks!

Reply:
It's not the same. Bash doesn't have remote machine to retrieve from.
You can check the equivalent Google/AWS operators. I don't see in any of them the functionality you ask to set here.

https://airflow.apache.org/docs/apache-airflow/stable/howto/define-extra-link.html

Reply:
My goal isn't to look at the logs. My goal is to get the last line from standard out and use it to determine if process succeeded or not.  This is exactly what the docker operator provides:
https://github.com/apache/airflow/blob/main/airflow/providers/docker/operators/docker.py#L142

Are you saying providing something similar to the way the docker operator works is the wrong approach?

Reply:
I agree this might be a useful feature actually - especially if it is an opt-in. I don't see a fundamental problem with it - generally speaking - every operator might decide what it wants to push to XCom. I am a little worried though about pulling all logs even if you want to push just one last line - unless there are are some smart generators implemented by the hook/library,  this one might be cause huge memory issues when the logs produced are large.

Is there any way you could optimize it ?

Reply:
This will be opt-in. It does not change the default logic or response at all.

The way the docker operator works, if you pass xcom_all = False only the last line to standard out is placed in XCOM, but if xcom_all=True then all of standard out is put in XCOM.

In this PR, I use the same logic as the docker operator but default the argument to NONE. If xcom_all is NONE, then nothing is put in XCOM. Assuming people don't set xcom_all, then this is a no-op (well, one if runs).

You can see the logic controlling this around line 316 of the operator in the PR.

Reply:
> You can see the logic controlling this around line 316 of the operator in the PR.

Yes I see `[-1]`. But does it mean that the whole log is kept in memory before doing that operation? Or is it somehow optimized under the hood to retrieve only last line when it is needed ? I am afraid that if you have 1GB log (which is not uncommon) you will download the whole log from remote service to **just** print one line. This is not only slow and will delay completion of the task a lot (sometimes minutes) in a way that is quite unexpected, but if the log is entirely loaded in memory, it means that the task will grow the memory by at least 1 GB - only for the purpose of pushing one line to XCo,mm.

Do you know if this is - or can be - optimized @perry2of5 ?

Reply:
I see what you mean. Reading the existing azure container instance operator code, it looks like _monitor_logging repeatedly pulls the entire log into memory and then logs just the new messages in _log_last.

https://github.com/apache/airflow/blob/main/airflow/providers/microsoft/azure/operators/container_instances.py#L352
https://github.com/apache/airflow/blob/main/airflow/providers/microsoft/azure/operators/container_instances.py#L387

Based on this, the operator is _already_ pulling the entire log into memory repeatedly. My modification doesn't make things worse, but it doesn't make things better either. I'd have to think about how we could optimize this.

It seems like my change isn't making anything worse. Faint praise....

Reply:
Yeah - in this case indeed it looks like "no worse", so if @eladkal has nothing against it, I am fine with getting that one in.

Reply:
Do you wish to weigh in @eladkal? If not, I'll close the conversation.